# Velocity adding in special relativity

## What happens when you add two speeds, each close to light speed?

Two objects are passing each other, each travelling at 99.99% $c$ (light speed) in opposite directions. What velocity do they measure for each other?

**99.99% $c$ + 99.99% $c$ = ?**

More than light speed? No, heres why.

## In everyday life

In everyday life, we can just add velocities like this:

\begin{equation}
v_{\text{total}} = v_1 + v_2
\end{equation}

So if an object travelling at 100 metres per second passes another object going at 100 m/s in the opposite direction, they measure each other's speed as just:

\begin{equation}
100 + 100 = 200
\end{equation}

But this common sense approach is only an approximation, and breaks down as speeds get closer to $c$.

## Relativistic velocity addition

The correct way to add any velocities is like this:

\begin{equation}
v_{\text{total}} = \frac{v_1 + v_2}{1 + \frac{v_1 v_2}{c^2}}
\end{equation}

The difference is microscopic for everyday speeds, so no one notices. But as you get closer to $c$ it becomes more significant.

The new equation is a bit more complicated, so let's write some Python code to help us in this Jupyter Notebook.

*You can download this notebook, and experiment with it on your PC or device. Try out your own velocity additions if you like!*

In [17]:
import math

c = 299_792_458.0  # speed of light, metres per second
c_squared = c**2  # c squared


def add_vel(v1: float, v2: float) -> float:
    if v1 >= c or v2 >= c:
        return math.nan
    return (v1 + v2) / (1.0 + (v1 * v2) / c_squared)


# test it
print(f"100 + 100 = {add_vel(100, 100)}")

100 + 100 = 199.99999999997775


Almost the same answer, to within 10 decimal places. Let's try it with some faster speeds.

In [18]:
from prettytable import PrettyTable, TableStyle
from IPython.display import Markdown

table = PrettyTable()
table.set_style(TableStyle.MARKDOWN)
table.align = "r"
table.field_names = [
    "Velocity1 (m/s)",
    "Velocity2 (m/s)",
    "Added relativistically (m/s)",
    "Discrepancy",
]


def calculate(v: float, round: bool = True) -> None:
    simple = v + v
    relativistic = add_vel(v, v)
    diff = 1.0 - relativistic / simple  # difference as a percentage
    if round:
        rel_str = f"{relativistic:,.1f}"
    else:
        rel_str = f"{relativistic:,}"
    table.add_row([f"{v:,}", f"{v:,}", rel_str, f"{diff:%}"])


# change these, but make sure they are less than light speed
calculate(1_000, False)
calculate(100_000, False)
calculate(1_000_000)
calculate(10_000_000)
calculate(100_000_000)
calculate(200_000_000)

print(f"Adding velocities relativistically, and discrepancy with newtonian physics")
markdown_table = table.get_string()
display(Markdown(markdown_table))

Adding velocities relativistically, and discrepancy with newtonian physics


| Velocity1 (m/s) | Velocity2 (m/s) | Added relativistically (m/s) | Discrepancy |
|---------------: |---------------: |----------------------------: |-----------: |
|           1,000 |           1,000 |          1,999.9999999777472 |   0.000000% |
|         100,000 |         100,000 |          199,999.97774700134 |   0.000011% |
|       1,000,000 |       1,000,000 |                  1,999,977.7 |   0.001113% |
|      10,000,000 |      10,000,000 |                 19,977,771.7 |   0.111141% |
|     100,000,000 |     100,000,000 |                179,975,072.5 |  10.012464% |
|     200,000,000 |     200,000,000 |                276,805,111.1 |  30.798722% |

## Effect gets proportionally larger

The difference between simple and relativistic addition gets larger as speeds increase. At 2 km/s it is almost zero, but at 400,000 km/s it has shot up to over 30%.

For faster speeds, let's switch to using fractions of $c$ (light speed) instead of m/s. Everything else is the same:

In [19]:
table = PrettyTable()
table.set_style(TableStyle.MARKDOWN)
table.align = "r"
table.field_names = [
    "Velocity1 (c)",
    "Velocity2 (c)",
    "Added relativistically (c)",
    "Discrepancy",
]


def add_vel_c(v1: float, v2: float) -> float:
    # two velocities as fraction of c
    if v1 >= 1.0 or v2 >= 1.0:
        return math.nan
    return (v1 + v2) / (1.0 + (v1 * v2))


def calculate_c(v: float) -> None:
    simple = v + v
    relativistic = add_vel_c(v, v)
    diff = 1.0 - relativistic / simple  # difference as a percentage
    table.add_row([f"{v:,}", f"{v:,}", f"{relativistic:,}", f"{diff:.1%}"])


# change these, but make sure they are less than light speed, 1.0
calculate_c(0.1)  # 10% light speed
calculate_c(0.2)
calculate_c(0.3)
calculate_c(0.4)
calculate_c(0.5)  # 50% light speed
calculate_c(0.7)
calculate_c(0.9)
calculate_c(0.95)
calculate_c(0.9999)
calculate_c(0.999999)
# calculate_c(0.999999999) # precision failure in normal Python
# calculate_c(1.0)

print(f"Adding velocities relativistically, and discrepancy with newtonian physics")
markdown_table = table.get_string()
display(Markdown(markdown_table))

Adding velocities relativistically, and discrepancy with newtonian physics


| Velocity1 (c) | Velocity2 (c) | Added relativistically (c) | Discrepancy |
|-------------: |-------------: |--------------------------: |-----------: |
|           0.1 |           0.1 |        0.19801980198019803 |        1.0% |
|           0.2 |           0.2 |        0.38461538461538464 |        3.8% |
|           0.3 |           0.3 |         0.5504587155963302 |        8.3% |
|           0.4 |           0.4 |          0.689655172413793 |       13.8% |
|           0.5 |           0.5 |                        0.8 |       20.0% |
|           0.7 |           0.7 |         0.9395973154362416 |       32.9% |
|           0.9 |           0.9 |          0.994475138121547 |       44.8% |
|          0.95 |          0.95 |         0.9986859395532195 |       47.4% |
|        0.9999 |        0.9999 |            0.9999999949995 |       50.0% |
|      0.999999 |      0.999999 |         0.9999999999995001 |       50.0% |

Notice how the discrepancy grows up to 50%, and the added velocities never exceed $c$.

For speeds of 0.999999999c and above we need more precision than ordinary Python can offer. So let's briefly switch to my Python library for *mpmath* solutions, that can handle hundreds of decimal places.

## Using the helper library for more precision

In [20]:
import mpmath as mp
import relativity_lib as rl

# Use 100 decimal places. Standard Python can only handle about 15-17 decimal places
rl.configure(100)

very_fast = mp.mpf("0.999999999") * rl.c
immensely_fast = mp.mpf("0.999999999999") * rl.c

builtin = add_vel_c(0.999999999, 0.999999999)
very_fast_2 = rl.add_velocities(very_fast, very_fast) / rl.c
immensely_fast_2 = rl.add_velocities(immensely_fast, immensely_fast) / rl.c

# compare built-in and library results
# print(f"Normal Python, precision failure: 0.999999999c * 2 = {builtin} ERROR!")
print(f"Helper library: 0.999999999c * 2 = {mp.nstr(very_fast_2, 20)}")
print(f"Helper library: 0.999999999999c * 2 = {mp.nstr(immensely_fast_2, 30)}")

Helper library: 0.999999999c * 2 = 0.9999999999999999995
Helper library: 0.999999999999c * 2 = 0.9999999999999999999999995


# So the answer to our question

**99.99% $c$ + 99.99% $c$ = 99.9999995% $c$**

How about that! Regardless of your velocities, you will never measure a speed greater than $c$.

## More info

For more on Special Relativity, check out 

[My Python relativity tools](https://github.com/lookbusy1344/Relativity/blob/main/Python/README.md)

Here's the wikipedia page on the topic:

https://en.wikipedia.org/wiki/Velocity-addition_formula#Special_relativity